### ISSUE TITLE
Deterministic module IDs are unstable in the face of change, which breaks long-term caching

### ISSUE THREAD

**Issue by robatwilliams** [webpack/webpack#14521]
# Bug report

**What is the current behavior?**
Adding extra imports of a module (or removing some) changes its module id, which causes a change in every bundle/chunk that uses that module.

I'm guessing this happens because occurrence order and/or frequency (for efficiency) contributes to the module id.

Example A: I add a new import of one of our existing app files. That module now gets a different module ID, so all 10+ bundles that use this module must change to reflect that.

Example B: I remove some imports of a library we use. It now gets a different module ID, so the vendor bundle changes with the new ID, plus every app code bundle that uses it.

**If the current behavior is a bug, please provide the steps to reproduce.**
Add or remove an import of a module.

**What is the expected behavior?**
There exists an id mechanism that is stable in the face of change.

**Other relevant information:**
webpack version:  5.48.0
Node.js version: 14.15.5
Operating System: Windows 10
Additional tools: n/a

**Comment by stanleyxu2005** [comment]
I have similar issue and narrow down to moduleId in Webpack 5, and then see this issue.

I rebuilt my project (source code not changed), after upgrading devDependencies (see fig. 1). 
All js files have different filenames (hash), but all css files are not changed.
So I compared both old and new built files, module id are changed. 

It seems not to be long-term caching. 
I'm looking for some help to see if I configured something wrong. :)

![fig. 1 - devDependencies change](https://github.com/webpack/webpack/assets/5162687/e529797f-ec35-4248-9578-0db7fa9d2a63)

![fig. 2 - js file diff](https://github.com/webpack/webpack/assets/5162687/7d087c84-f22d-49c9-aab8-7883e5fdd66e)

![fig. 3 - webpack config](https://github.com/webpack/webpack/assets/5162687/a0a4d5df-c584-4641-85e3-4f835988c746)

**Comment by alexander-akait** [comment]
Did you update webpack? If yes - expected, sometimes we change some logic so you can have different IDs between webpack version, we don't do it often, but sometimes it happens

**Comment by stanleyxu2005** [comment]
> Did you update webpack? If yes - expected, sometimes we change some logic so you can have different IDs between webpack version, we don't do it often, but sometimes it happens

Hi Alex.

No. If I update any of the devDependencies (babel, or loader), it will invalidate all filenames....
I tested one by one. 
But if I add a new devDependency, which does not have anything to do with Webpack, nothing changes. (e.g. d3)

**Comment by alexander-akait** [comment]
Can you provide small reproducible example?

**Comment by stanleyxu2005** [comment]
> Can you provide small reproducible example?

Dear Lord, here is the repo for a test.
https://github.com/stanleyxu2005/webpack-unstable-moduleid

For the time being, I test with one dev dep `css-loader`. The source code does not have any css code at all, but after upgrade/downgrade `css-loader`, all built JS filenames will be changed.

_The webpack config is a bit messy. Maybe the issue is caused by my wrong usage._

**Comment by alexander-akait** [comment]
@stanleyxu2005 Weird, I can't reproduce (linux ubuntu)... Do you use windows? Can you try to disable minimizer (terser) and look at diff with `"css-loader": "=6.9.1"` and `"css-loader": "=6.9.0"` and show me differents?

**Comment by stanleyxu2005** [comment]
> minimizer
```
      // new TerserPlugin({
      //   extractComments: {
      //     // Precisely extract vendor licenses
      //     condition: /^\**!|LICENSE|@preserve|@cc_on/i,
      //   },
      // }),
```
OS: Windows 11
Node: v18.17.1

![image](https://github.com/webpack/webpack/assets/5162687/cbdd5541-a108-47d9-a409-f5572091d856)
![image](https://github.com/webpack/webpack/assets/5162687/c00e2989-6eba-4346-ae10-73b92db4523a)

**Comment by alexander-akait** [comment]
@stanleyxu2005 Yeah, I see it, so because you have different content you have different hashes, we are calculating hashes before minification, in future we want to improve it, anyway even without minimizer you will have different hashes...

> It's more or less related to pnpm, due to its special way to manage dependencies.

It is possible to avoid such behaviour (some flags or setting), like we have lock files?

**Comment by stanleyxu2005** [comment]
> @stanleyxu2005 Yeah, I see it, so because you have different content you have different hashes, we are calculating hashes before minification, in future we want to improve it, anyway even without minimizer you will have different hashes...
👍 
> 
> > It's more or less related to pnpm, due to its special way to manage dependencies.
> 
> It is possible to avoid such behaviour (some flags or setting), like we have lock files?

Yes. I have `pnpm-lock.yaml`. Now it's clear, unless I have a good reason, I should not upgrade dev deps. It will invalidate long-term cache.

**Comment by alexander-akait** [comment]
@stanleyxu2005 You can have `pnpm-lock.yaml` and use bots to make updates (or make it manually once a week for example), also you can setup basic tests so CI will allow to test your build

**Comment by stanleyxu2005** [comment]
> @stanleyxu2005 You can have `pnpm-lock.yaml` and use bots to make updates (or make it manually once a week for example), also you can setup basic tests so CI will allow to test your build

I do have `pnpm-lock.yaml`. 

As long as I don't upgrade any dev dependencies, the filename and file contents of production build are stable.

The problem is if I want to upgrade a dev dependencies, e.g. changing webpack v5.90.0 to v5.91.0, then all filenames and file contents of production build will be changed. (When user launches webapp, all files have to be redownloaded, because js/css filenames are changed, although there is no actual change in the app.)

You helped to analyse the issue before 
https://github.com/webpack/webpack/issues/14521#issuecomment-1925319751
https://github.com/webpack/webpack/issues/14521#issuecomment-1925313621

You mentioned webpack will consider to improve that. I'd like to know, if this will be planed in Webpack v5 or in future major version? And if I'd like to upgrade dev deps regularly, is there a workaround there? 

Thanks

**Comment by stanleyxu2005** [comment]
> > @stanleyxu2005 You can have `pnpm-lock.yaml` and use bots to make updates (or make it manually once a week for example), also you can setup basic tests so CI will allow to test your build
> 
> I do have `pnpm-lock.yaml`.
> 
> As long as I don't upgrade any dev dependencies, the filename and file contents of production build are stable.
> 
> The problem is if I want to upgrade a dev dependencies, e.g. changing webpack v5.90.0 to v5.91.0, then all filenames and file contents of production build will be changed.
> 
> You helped to analyse the issue before [#14521 (comment)](https://github.com/webpack/webpack/issues/14521#issuecomment-1925319751) [#14521 (comment)](https://github.com/webpack/webpack/issues/14521#issuecomment-1925313621)
> 
> You mentioned webpack will consider to improve that. I'd like to know, if this will be planed in Webpack v5 or in future major version? And if I'd like to upgrade dev deps regularly, is there a workaround there?
> 
> Thanks

@alexander-akait I managed to solve my problem with this workaround for my case
 
![image](https://github.com/webpack/webpack/assets/5162687/c1ef2b57-54d2-457e-916f-fce7c076c412)
